lib/datadog/di/probe_notification_builder.rb

```diff
@@ -0,0 +1,212 @@
+          status: 'RECEIVED',)
+      end
+
+      def build_installed(probe)
+        build_status(probe,
+          message: "Probe #{probe.id} has been instrumented correctly",
+          status: 'INSTALLED',)
+      end
+
+      def build_emitting(probe)
+        build_status(probe,
+          message: "Probe #{probe.id} is emitting",
+          status: 'EMITTING',)
+      end
+
+      def build_executed(probe,
+        trace_point: nil, rv: nil, duration: nil, caller_locations: nil,
+        args: nil, kwargs: nil, serialized_entry_args: nil)
+        snapshot = if probe.line? && probe.capture_snapshot?
+          if trace_point.nil?
+            raise "Cannot create snapshot because there is no trace point"
+          end
+          get_local_variables(trace_point)
+        end
+        # TODO check how many stack frames we should be keeping/sending,
+        # this should be all frames for enriched probes and no frames for
+        # non-enriched probes?
+        build_snapshot(probe, rv: rv, snapshot: snapshot,
+          duration: duration, caller_locations: caller_locations, args: args, kwargs: kwargs,
+          serialized_entry_args: serialized_entry_args)
+      end
+
+      def build_snapshot(probe, rv: nil, snapshot: nil,
+        duration: nil, caller_locations: nil, args: nil, kwargs: nil,
+        serialized_entry_args: nil)
+        # TODO also verify that non-capturing probe does not pass
+        # snapshot or vars/args into this method
+        captures = if probe.capture_snapshot?
+          if probe.method?
+            {
+              entry: {
+                # standard:disable all
+                arguments: if serialized_entry_args
+                  serialized_entry_args
+                else
+                  (args || kwargs) && serializer.serialize_args(args, kwargs)
+                end,
+                throwable: nil,
+                # standard:enable all
+              },
+              return: {
+                arguments: {
+                  "@return": serializer.serialize_value(rv),
+                },
+                throwable: nil,
+              },
+            }
+          elsif probe.line?
+            {
+              lines: snapshot && {
+                probe.line_no => {locals: serializer.serialize_vars(snapshot)},
+              },
+            }
+          end
+        end
+
+        location = if probe.line?
+          actual_file = if probe.file
+            # Normally caller_locations should always be filled for a line probe
+            # but in the test suite we don't always provide all arguments.
+            actual_file_basename = File.basename(probe.file)
+            caller_locations&.detect do |loc|
+              # TODO record actual path that probe was installed into,
+              # perform exact match here against that path.
+              File.basename(loc.path) == actual_file_basename
+            end&.path || probe.file
```

Comment:
This seems like a slightly expensive thing to do for just the tests :thinking:. Or can there be situations in user code where this gets triggered?

Reply:
We need the path to the file in the payload, but the path should be determined during instrumentation as per the comment just above. I've hacked path matching and processing by doing basename comparisons but this isn't sufficient, for v1 I will have the actual path stored in the probe after instrumentation and nothing should be calculated here.